### PR TITLE
return 400 instead of 500 for requests with invalid URI scheme #36

### LINKED ISSUE #36: Invalid request returns 500 instead of 400 in akka http server
<a href="https://github.com/francisdb"></img></a> **Issue by [francisdb](https://github.com/francisdb)**
_Wednesday Aug 24, 2016 at 08:26 GMT_
_Originally opened as https://github.com/akka/akka/issues/21272_

---

We got this on akka http 2.4.6

Not sure how a request with invalid uri ends up at the server but I suppose this should not be a 500 error but rather a 400?

```
Internal server error, sending 500 response

java.lang.IllegalArgumentException: `uri` must have scheme "http", "https" or no scheme
    at akka.http.scaladsl.model.HttpRequest$.verifyUri(HttpMessage.scala:292)
    at akka.http.scaladsl.model.HttpRequest.<init>(HttpMessage.scala:155)
    at akka.http.scaladsl.model.HttpRequest$.apply(HttpMessage.scala:302)
    at akka.http.impl.engine.server.HttpServerBluePrint$PrepareRequests$$anon$1.onPush(HttpServerBluePrint.scala:133)
    at akka.stream.impl.fusing.GraphInterpreter.processElement$1(GraphInterpreter.scala:587)
    at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:598)
    at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:539)
    at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:472)
    at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:424)
    at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:604)
    at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:619)
    at akka.actor.Actor$class.aroundReceive(Actor.scala:482)
    at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:530)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
    at akka.actor.ActorCell.invoke(ActorCell.scala:495)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
    at akka.dispatch.Mailbox.run(Mailbox.scala:224)
    at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

### PR DESCRIPTION
**Pull request by shkoder**
fixes #36

### CODE REVIEW

File: akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerSpec.scala [hunk 1]
```diff
@@ -874,9 +874,9 @@ class HttpServerSpec extends AkkaSpec(
            |Date: XXXX
            |Connection: close
            |Content-Type: text/plain; charset=UTF-8
-           |Content-Length: 41
+           |Content-Length: 167
            |
-           |Request is missing required `Host` header""")
+           |Cannot establish effective URI of request to `/abc`, request has a relative URI and is missing a `Host` header; consider setting `akka.http.server.default-host-header`""")
```
[the remark this answers is not in the document]
Comment (shkoder) [reply akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerSpec.scala]: good idea, addressed with f3abbe1

File: akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerSpec.scala [hunk 2]
```diff
@@ -1290,6 +1290,50 @@ class HttpServerSpec extends AkkaSpec(
...
+           |Content-Type: text/plain; charset=UTF-8
+           |Content-Length: 64
+           |
+           |`uri` must have scheme "http", "https", "ws", "wss" or no scheme""")
+
+      netIn.sendComplete()
+      netOut.expectComplete()
+    })
+
+    "reject HTTP/1.1 requests with invalid Host header schema" in assertAllStagesStopped(new TestSetup {
```
Comment (jrudolph) [inline akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerSpec.scala]: The name of the test seems slightly off.

`"reject HTTP/1.1 requests with Host header that doesn't match absolute request target authority"`

File: akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerSpec.scala [hunk 3]
```diff
@@ -874,9 +874,9 @@ class HttpServerSpec extends AkkaSpec(
            |Date: XXXX
            |Connection: close
            |Content-Type: text/plain; charset=UTF-8
-           |Content-Length: 41
+           |Content-Length: 110
            |
-           |Request is missing required `Host` header""")
+           |Cannot establish effective URI of request to `/abc`, request has a relative URI and is missing a `Host` header""")
```
Comment (jrudolph) [inline akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerSpec.scala]: :+1:

File: akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala [hunk 4]
```diff
@@ -216,10 +216,11 @@ private[http] object HttpServerBluePrint {
       case start: RequestStart ⇒
         try {
           val effectiveUri = HttpRequest.effectiveUri(start.uri, start.headers, isSecureConnection, defaultHostHeader)
+          HttpRequest.verifyUri(effectiveUri)
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala]: If we verify it here we might want to remove the other check in `HttpRequest` for performance reasons. Or, we don't verify here and catch and handle the exception at the place where the request is created. I guess handling it at the place where the HttpRequest is created would be simpler to achieve.

**Comment by shkoder** [comment]
@jrudolph I've removed the double verification with 5b4c75d

**Comment by jrudolph** [comment]
Great stuff, thanks a lot, @shkoder!

**Pull request merged**